File: opt_four_checker.cpp
#include<iostream>
#include<vector>
#include<utility>
#include<tuple>
#include<map>
#include<string>

using namespace std;

int dist(pair<int, int> x, pair<int, int> y){
    return abs(x.first - y.first)+abs(x.second - y.second);
}

int distf(pair<int, int> x, pair<int, int> y, pair<int, int> x1, pair<int, int> y1){
    return min(dist(x, y), min(dist(x, x1)+dist(y1, y)+1, dist(x, y1)+dist(x1, y)+1));
}

bool corner_check(pair<int, int> x, int n, int m){
    pair<int, int> p(1, 1), q(1, m), r(n, m), s(n, 1);
    return x != p && x != q && x != r && x != s;
}

bool four_cond(pair<int, int> x1, pair<int, int> y1, int n, int m){
    int Gain1 = abs(abs(x1.second - y1.second) - abs(x1.first - y1.first)) - 1;
    
    bool first = corner_check(x1, n, m) && corner_check(y1, n, m);
    bool second = (Gain1 % 2 == 0) && (Gain1 > 0);
    bool third = 2 * min(abs(y1.first - x1.first), abs(x1.second - y1.second)) >= (Gain1 + 4);
    return first && second && third;
}

int main(int argc, char* argv[]){
    if(argc < 3){
        cout<<"Please provide dimensions of the grid.\n";
        exit(-1);
    }
    int n = stoi(argv[1]);
    int m = stoi(argv[2]);
    vector< pair<int, int> > points, border_points;

    for (int i = 1; i <= n; i++){
        for (int j = 1; j <= m; j++){
            pair<int, int> p(i, j);
            points.push_back(p);
            if(i == 1 || j == 1 || i == n || j == m)
                border_points.push_back(p);
        }
    }
    cout<<border_points.size()<<endl;

    vector< pair< pair<int, int>, pair<int, int> > > point_pairs;

    for (int i = 0; i < points.size(); i++){
        for (int j = i+1; j < points.size(); j++){
            pair< pair<int, int>, pair<int, int> > p(points[i], points[j]);
            point_pairs.push_back(p);
        }
    }

    vector< tuple<pair<int, int>, pair<int, int>, pair<int, int> > > three_border_points;

    for (int i = 0; i < border_points.size(); i++){
        for (int j = i+1; j < border_points.size(); j++){
            for(int k = j+1; k < border_points.size(); k++){
                tuple< pair<int, int>, pair<int, int>, pair<int, int> > tup;
                tup = make_tuple(border_points[i], border_points[j], border_points[k]);
                three_border_points.push_back(tup);
            }
        }
    }
            
    for (int i = 0; i < point_pairs.size(); i++){
        pair<int, int> x = point_pairs[i].first, y = point_pairs[i].second;
        if(dist(x, y) > 1 && (x.first <= y.first) && (x.second >= y.second) && (x.second - y.second) >= (y.first - x.first)){
            bool found = true;
            for(int j = 0; j < three_border_points.size(); j++){
                map< tuple <int, int, int>, int> distances;
                bool flag = true;
                for (int t = 0; t < points.size(); t++){
                    tuple<int, int, int> tup;
                    tup = make_tuple(distf(points[t], get<0>(three_border_points[j]), x, y), distf(points[t], get<1>(three_border_points[j]), x, y), distf(points[t], get<2>(three_border_points[j]), x, y));
                    if(distances.count(tup) > 0){
                        flag = false;
                        break;
                    }
                    else
                        distances[tup] = 1;
                }
                if(flag){
                    found = false;
                    break;
                }   
            }
            bool cond = four_cond(x, y, n, m);
            if(!(cond ^ found)){
                if(found)
                    cout<<"MD is 4 when edge is between ("<<x.first<<","<<x.second<<") and ("<<y.first<<","<<y.second<<")\n";
            }
            else{
                cout<<"Mistake\n";
                exit(-1);
            }
        }
    }
    cout<<"Success!\n";
}

File: python/opt_four_checker.py
from utils import dist, distf, four_checker

import itertools
import pickle
import sys
import time

if len(sys.argv) < 3:
	print("Please provide dimensions of the grid.")
	exit(1)
n = int(sys.argv[1])
m = int(sys.argv[2])

t0 = time.time()
point_list = itertools.product(range(1, n+1), range(1, m+1))
point_list = list(point_list)

border_list = []

for i in range(1, n+1):
	border_list.append((i, 1))
	border_list.append((i, m))

for i in range(1, m+1):
	border_list.append((1, i))
	border_list.append((n, i))

border_list = list(set(border_list))

for ((x1, y1), (x2, y2)) in itertools.combinations(point_list, 2):	
	if dist(x1, y1, x2, y2) > 1 and (x1 <= x2) and (y1 >= y2) and (y1 - y2) >= (x2 - x1):
		
		found = True

		for ((a1, b1), (a2, b2), (a3, b3)) in itertools.combinations(border_list, 3):
			distances = {}
			flag = True
	
			for (i,j) in point_list:

				tup = (distf(i, j, a1, b1, x1, y1, x2, y2), distf(i, j, a2, b2, x1, y1, x2, y2), distf(i, j, a3, b3, x1, y1, x2, y2))
				if tup in distances:
					flag = False
					break
				else:
					distances[tup] = 1
			if flag == True:
				found = False
				break
		cond = four_checker(x1, y1, x2, y2, n, m)

		if found ^ cond == False:
			if found == True:
				print("MD is 4 when edge is between", str((x1, y1)), "and", str((x2, y2)))
		else:
			print("Mistake in ", x1, y1, x2, y2)
			exit()
		
print("Success. Conjecture for", n, "X", m, "grid is verified when MD is 4.")
t1 = time.time()		
print("Time taken: ", t1 - t0)